pallets/pallet-asset-switch/src/lib.rs

```diff
@@ -0,0 +1,686 @@
+			crate::try_state::do_try_state::<T, I>(n)
+		}
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config<I>, I: 'static = ()> {
+		/// A new switch pair is created.
+		SwitchPairCreated {
+			remote_asset_circulating_supply: u128,
+			remote_asset_ed: u128,
+			pool_account: T::AccountId,
+			remote_asset_id: VersionedAssetId,
+			remote_reserve_location: VersionedLocation,
+			remote_xcm_fee: Box<VersionedAsset>,
+			remote_asset_total_supply: u128,
+		},
+		/// A switch pair is removed.
+		SwitchPairRemoved { remote_asset_id: VersionedAssetId },
+		/// A switch pair has enabled switches.
+		SwitchPairResumed { remote_asset_id: VersionedAssetId },
+		/// A switch pair has suspended switches.
+		SwitchPairPaused { remote_asset_id: VersionedAssetId },
+		/// The XCM fee for the switch has been updated.
+		SwitchPairFeeUpdated { old: VersionedAsset, new: VersionedAsset },
+		/// A switch of local -> remote asset has taken place.
+		LocalToRemoteSwitchExecuted {
+			from: T::AccountId,
+			to: VersionedLocation,
+			amount: LocalCurrencyBalanceOf<T, I>,
+		},
+		/// A switch of remote -> local asset has taken place.
+		RemoteToLocalSwitchExecuted { to: T::AccountId, amount: u128 },
+	}
+
+	#[pallet::error]
+	pub enum Error<T, I = ()> {
+		/// Provided switch pair info is not valid.
+		InvalidInput,
+		/// The runtime-injected logic returned an error with a specific code.
+		Hook(u8),
+		/// There are not enough remote assets to cover the specified amount of
+		/// local tokens to switch.
+		Liquidity,
+		/// Failure in transferring the local tokens from the user's balance to
+		/// the switch pair pool account.
+		LocalPoolBalance,
+		/// The calculated switch pair pool account does not have enough local
+		/// tokens to cover the specified `circulating_supply`.
+		PoolInitialLiquidityRequirement,
+		/// A switch pair has already been set.
+		SwitchPairAlreadyExisting,
+		/// The switch pair did not enable switches.
+		SwitchPairNotEnabled,
+		/// No switch pair found.
+		SwitchPairNotFound,
+		/// The user does not have enough local tokens to cover the requested
+		/// switch.
+		UserSwitchBalance,
+		/// The user does not have enough assets to pay for the remote XCM fees.
+		UserXcmBalance,
+		/// Something regarding XCM went wrong.
+		Xcm,
+		/// Internal error.
+		Internal,
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn switch_pair)]
+	pub(crate) type SwitchPair<T: Config<I>, I: 'static = ()> = StorageValue<_, SwitchPairInfoOf<T>, OptionQuery>;
+
+	#[pallet::call]
+	impl<T: Config<I>, I: 'static> Pallet<T, I>
+	where
+		LocalCurrencyBalanceOf<T, I>: Into<u128>,
+	{
+		/// Set a new switch pair.
+		///
+		/// See the crate's README for more.
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config<I>>::WeightInfo::set_switch_pair())]
+		pub fn set_switch_pair(
+			origin: OriginFor<T>,
+			remote_asset_total_supply: u128,
```

Comment:
Yes but we can't do anything about it. Also the fungibility of XCM assets uses `u128`. So that's the maximum that we can safely assume can be represented over XCM.

Reply:
Ah yeah true. Fungible is set to u128